File: src/pnp/compiler/model/expression/operation/UnaryOperation.java
package pnp.compiler.model.expression.operation;

import pnp.compiler.model.expression.Expression;
import pnp.compiler.model.construct.type.Type;

public class UnaryOperation implements Operation {
    public Expression operand;
    public final Operator operator;
    public Type resultType;
    public Type operandType;

    public UnaryOperation(Operator operator, Expression operand, Type resultType) {
        this.operator = operator;
        this.operand = operand;
        this.resultType = resultType;
        this.operandType = resultType;
    }

    public UnaryOperation(Operator operator, Expression operand, Type resultType, Type operandType) {
        this.operator = operator;
        this.operand = operand;
        this.resultType = resultType;
        this.operandType = operandType;
    }

    @Override
    public Type getType() {
        return resultType;
    }
}

File: src/main/kotlin/br/com/pnp/model/expression/operation/UnaryOperation.kt
package br.com.pnp.model.expression.operation

import br.com.pnp.model.construct.type.Type
import br.com.pnp.model.expression.Expression

class UnaryOperation(
    val operator: Operator,
    var operand: Expression,
    resultType: Type,
    var operandType: Type = resultType
) : Operation {
    override val type: Type = resultType
}
